testserver/testserver.go

```diff
@@ -0,0 +1,303 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package testserver
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+	"syscall"
+)
+
+var sqlURLRegexp = regexp.MustCompile("sql:\\s+(postgresql:.+)\n")
+
+const (
+	stateNew     = iota
+	stateRunning = iota
+	stateStopped = iota
+	stateFailed  = iota
+)
+
+// TestServer is a helper to run a real cockroach node.
+type TestServer struct {
+	mu        sync.RWMutex
+	state     int
+	baseDir   string
+	Cmd       *exec.Cmd
+	Error     error
+	Args      []string
+	stdout    string
+	stderr    string
+	stdoutBuf logWriter
+	stderrBuf logWriter
+}
+
+// NewTestServer creates a new TestServer, but does not start it.
+// The cockroach binary for your OS and ARCH is downloaded automatically.
+// If the download fails, we attempt just call "cockroach", hoping it is
+// found in your path.
+func NewTestServer() (*TestServer, error) {
+	cockroachBinary, err := downloadLatestBinary()
+	if err == nil {
+		log.Printf("Using automatically-downloaded binary: %s", cockroachBinary)
+	} else {
+		log.Printf("Attempting to use cockroach binary from your PATH")
+		cockroachBinary = "cockroach"
+	}
+
+	baseDir, err := ioutil.TempDir("", "cockroach-testserver")
+	if err != nil {
+		return nil, fmt.Errorf("could not create temp directory: %s", err)
+	}
+
+	logDir := filepath.Join(baseDir, "logs")
+	if err := os.MkdirAll(logDir, 0755); err != nil {
+		return nil, fmt.Errorf("could not create logs directory: %s: %s", logDir, err)
+	}
+
+	args := []string{
+		cockroachBinary,
+		"start",
+		"--logtostderr",
+		"--insecure",
+		"--port=0",
+		"--http-port=0",
+		"--store=" + baseDir,
+	}
+
+	ts := &TestServer{
+		baseDir: baseDir,
+		Args:    args,
+		stdout:  filepath.Join(logDir, "cockroach.stdout"),
+		stderr:  filepath.Join(logDir, "cockroach.stderr"),
+	}
+	return ts, nil
+}
+
+// Stdout returns the entire contents of the process' stdout.
+func (ts *TestServer) Stdout() string {
+	return ts.stdoutBuf.String()
+}
+
+// Stderr returns the entire contents of the process' stderr.
+func (ts *TestServer) Stderr() string {
+	return ts.stderrBuf.String()
+}
+
+// PGURL parses the node's stdout log for the postgres URL.
+// It returns an actual URL so that callers may modify individual settings.
+// eg:
+//   u := server.PGURL()
+//   u.Path = "mydatabase"
+//   db, err := sql.Open("postgres", u.String())
+// NOTE: it may not be available for a few milliseconds after start,
```

Comment:
Why add the additional method? Just simplify and have this one which should use a mutex to block and then cache the URL result once it's computed.